CIP-0032/README.md

```diff
@@ -115,14 +115,51 @@ However, the other approaches also have some advantages.
+There are reasons to want to do this: _not_ using inline datums causes communication issues for users, and so it is quite reasonable that an application developer may want to be able to enforce their use.
+
+Furthermore, as a general principle we try to keep the script context as faithful to real transactions as possible.
+Even if the use for the information is not immediately obvious, we try to err on the side of providing it and letting users decide.
+
+Hence we _do_ include information about inline datums in the script context.
+
+#### Representation and backwards compatibility
+
+There are a couple of options for how to change the representation of the script context to include the new information, and whether to make a backwards compatibility effort for old language versions.
+
+For the new script context:
+
+1. Match the ledger representation as much as possible: change the fields on inputs and outputs to be either a datum hash or a datum.
+2. Try to only have one way to look up datums: put inline datums in the datum witness map and insert their hashes into the corresponding inputs and outputs; optionally add a boolean to inputs and outputs to indicate whether the datum was originally inline.
+
+For backwards compatibility:
+
+1. Don't try to represent inline datums for scripts using old language versions: old scripts simply can't be run in transactions that use inline datums (since we can't represent the information).
+2. Rewrite inline datums as non-inline datums: put inline datums in the datum witness map and insert their hashes into the corresponding inputs and outputs.
+
+For the new script context, option 1 has the significant advantage of matching the ledger representation of transactions.
+This makes it easier to implement, and also avoids conceptual overhead for users who would have to distinguish the two ways of representing transactions.
+While the conceptual distance here may be small, if we let it grow over time then it may become quite confusing.
+
+We then have the choice of what to do about backwards compatibility.
+Option 2 would work, but is more complicated for the ledger and is inconsistent in representation with our choice for the new script context (inline datums are sometimes represented faithfully, and sometimes put in the witness map).
+Option 1 is simple, but doesn't allow old scripts to work with inline datums.
+This would not be so bad if it just meant that old scripts could not be spent in transactions that include inline datums, but it also introduces a new way to make an unspendable output.
+If a user creates an output with an old script and an inline datum, then any transaction spending that output will include an inline datum, which we would not allow.
+
+Unfortunately, we cannot prevent users from creating such outputs in general, since script addresses do not include the script language, so the ledger cannot tell whether the inline datum is permissible or not.
```

Comment:
Even if it did include this information, it would still not be enough. The script might be V2, but to consume it you might need another input that uses a V1 validator, or you might need to mint/burn with a V1 minting policy, which wouldn't work because those scripts can't be run.

Reply:
This could be because e.g. the datum of the UTXO in question contains a reference to a minting policy hash, and the script asserts that a token of that minting policy is minted. In general, this problem is only solvable by the client knowing the contents of the scripts in question, so I don't think there's any way around this.

Reply:
Right, that would be a more elaborate way of getting yourself stuck in this position indeed.